src/npcmove.cpp

```diff
@@ -3020,13 +3021,33 @@ void npc::set_destination()
     tripoint surface_omt_loc = global_omt_location();
     surface_omt_loc.z = 0;
 
-    std::string dest_type = get_location_for( needs.front() )->get_random_terrain().id().str();
+    std::string dest_type, current_dest_type;
+    tripoint current_goal;
+    int current_goal_dist, goal_dist = -1;
+    npc_need needs_final;
+    for( npc_need &i : needs ) {
+        for( int k = 0; k < 30; k++ ) {
+            current_dest_type = get_location_for( i )->get_random_terrain().id().str();
+            current_goal = overmap_buffer.find_closest( surface_omt_loc, current_dest_type, 0, false, false,
+                           visited_point );
+            current_goal_dist = square_dist( surface_omt_loc, current_goal );
+            if( ( current_goal_dist < goal_dist ) || ( goal_dist == -1 ) ) {
+                goal = current_goal;
+                goal_dist = current_goal_dist;
+                needs_final = i; // only for debug purpose
+                dest_type = current_dest_type; // same with needs_final
+            }
+        }
+    }
+    /*std::string dest_type = get_location_for( needs.front() )->get_random_terrain().id().str();
```

Comment:
Please do not check in commented out code.

Reply:
done!